modules/core/src/main/java/org/apache/ignite/internal/jdbc2/JdbcBinaryBuffer.java

```diff
@@ -0,0 +1,454 @@
+        this.len = len;
+        this.isReadOnly = isReadOnly;
+    }
+
+    /**
+     * Create shallow read-only copy of this buffer.
+     */
+    public JdbcBinaryBuffer shallowCopy() {
+        return new JdbcBinaryBuffer(arr, off, len, true);
+    }
+
+    /**
+     * Provide InputStream through which the data can be read starting from the
+     * begining.
+     *
+     * <p>Stream is not limited meaning that it would return any new data
+     * written to the buffer after stream creation.
+     *
+     * @return InputStream instance.
+     */
+    public InputStream getInputStream() {
+        return new BufferInputStream();
+    }
+
+    /**
+     * Get copy of the buffer data as byte array.
+     *
+     * @return Byte array containing buffer data.
+     */
+    public byte[] getBytes() {
+        byte[] bytes = new byte[len];
+
+        read(0, bytes, 0, len);
+
+        return bytes;
+    }
+
+    /**
+     * Provide OutputStream through which the data can be written to buffer starting from
+     * the (zero-based) {@code pos} position.
+     *
+     * @param pos The zero-based offset to the first byte to be written. Must not be negative
+     *            or greater than total count of bytes in buffer.
+     *
+     * @return OutputStream instance.
+     */
+    OutputStream getOutputStream(int pos) {
+        return new BufferOutputStream(pos);
+    }
+
+    /**
+     * Provide InputStream through which the no more than {@code len} bytes can be read
+     * from buffer starting from the specified zero-based position {@code pos}.
+     *
+     * @param pos The zero-based offset to the first byte to be retrieved. Must not be negative
+     *            or greater than total count of bytes in buffer.
+     * @param len The length in bytes of the data to be retrieved. Must not be negative.
+     * @return InputStream instance.
+     */
+    InputStream getInputStream(int pos, int len) {
+        return new BufferInputStream(pos, len);
+    }
+
+    /**
+     * Truncate data in this buffer to specified length.
+     *
+     * @param len New length.
+     */
+    void truncate(int len) {
+        this.len = len;
+
+        reallocate(Math.max(MIN_CAP, len));
+    }
+
+    /**
+     * @return Length of data in this buffer.
+     */
+    int length() {
+        return len;
+    }
+
+    /**
+     * Read up to {@code resLen} bytes from this buffer from specified position {@code pos}.
+     *
+     * @param pos Pointer to a position.
+     * @param resBuf Output byte array to write to.
+     * @param resOff Offset in the output array to start write to.
+     * @param resLen Number of bytes to read.
+     * @return Number of bytes read. -1 if end of data reached.
+     */
+    int read(int pos, byte[] resBuf, int resOff, int resLen) {
+        Objects.checkFromIndexSize(resOff, resLen, resBuf.length);
+
+        if (pos >= len)
+            return -1;
+
+        int bufOff = pos + off;
+
+        int size = Math.min(resLen, len - pos);
```

Comment:
`size` may be inlined.
`U.arrayCopy` returns number of written bytes. We can use it to return restul.

```
return U.arrayCopy(arr, pos + off, resBuf, resOff, Math.min(resLen, len - pos);
```

Reply:
`U.arrayCopy` doesn't return number of written bytes.

Instead it returns new offset in the result array:

```
public static int arrayCopy(byte[] src, int off, byte[] resBuf, int resOff, int len) {
    .....
    return resOff + len;
}
```